### PR TITLE
multiple finders

### PR DESCRIPTION
**Pull request by narora1**

### CODE REVIEW

File: robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h [hunk 1]
```diff
@@ -106,11 +106,21 @@ struct Camera3dToArmError
                                      CalibrationOffsetParser* offsets,
                                      robot_calibration_msgs::CalibrationData& data)
   {
+    int index = -1;
+      for (size_t k =0; k < data.observations.size() ; k++)
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h]: indentation of this block is off

Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h]: space also needed after k =0

File: robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h [hunk 2]
```diff
@@ -106,11 +106,21 @@ struct Camera3dToArmError
...
+      {
+        if ( data.observations[k].sensor_name == "camera")
+        {
+          index = k;
+          break;
+        }
+      }
+
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h]: need to add a check -- if index is -1 then we have to exit with error (rather than segfault)

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 3]
```diff
@@ -56,7 +56,7 @@ struct GroundPlaneError
     // Compute residuals
     for (size_t i = 0; i < camera_pts.size() ; ++i)
     {
-      residuals[i] = camera_pts[i].point.z - z_;  // if camera_pts is in base frame
+      residuals[i] = (camera_pts[i].point.z - z_) * (camera_pts[i].point.z - z_);  // if camera_pts is in base frame
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: this should be in a different commit -- it's a change in the actual residual

File: robot_calibration/src/led_finder.cpp [hunk 4]
```diff
@@ -56,9 +58,9 @@ LedFinder::LedFinder(ros::NodeHandle& nh) :
   // Setup subscriber
   nh.param<std::string>("topic", topic_name, "/points");
   subscriber_ = nh.subscribe(topic_name,
-                             1,
-                             &LedFinder::cameraCallback,
-                             this);
+                1,
```
Comment (mikeferguson) [inline robot_calibration/src/led_finder.cpp]: revert these changes -- the indentation is correct

File: robot_calibration/src/led_finder.cpp [hunk 5]
```diff
@@ -459,8 +492,8 @@ bool LedFinder::CloudDifferenceTracker::isFound(
 }
 
 bool LedFinder::CloudDifferenceTracker::getRefinedCentroid(
-  const sensor_msgs::PointCloud2& cloud,
-  geometry_msgs::PointStamped& centroid)
+    const sensor_msgs::PointCloud2& cloud,
+    geometry_msgs::PointStamped& centroid)
```
Comment (mikeferguson) [inline robot_calibration/src/led_finder.cpp]: revert these changes -- the indentation is correct

File: robot_calibration/src/optimizer.cpp [hunk 6]
```diff
@@ -146,32 +144,42 @@ int Optimizer::optimize(OptimizationParams& params,
...
-          for (size_t k = 0; k < data[i].observations[0].features.size(); ++k)
+          for (size_t k = 0; k < data[i].observations[index].features.size(); ++k)
             std::cout << "  " << std::setw(10) << std::fixed << residuals[(3*k + 2)];
           std::cout << std::endl << std::endl;
         }
 
         problem->AddResidualBlock(cost,
-                                   NULL /* squared loss */,
-                                   free_params);
+                                  NULL,//squared loss ,
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: need 2 spaces after , and one space after // (and remove extra comma end of line)

File: robot_calibration/src/optimizer.cpp [hunk 7]
```diff
@@ -188,16 +196,26 @@ int Optimizer::optimize(OptimizationParams& params,
           z_,
           offsets_.get(), data[i]);
 
+        int index = -1;
+        for (size_t k =0; k < data[i].observations.size() ; k++)
+        {
+            if ( data[i].observations[k].sensor_name == camera_name)
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: this block is indented too much -- and no space before data

File: robot_calibration/src/checkerboard_finder.cpp [hunk 8]
```diff
@@ -172,11 +172,41 @@ bool CheckerboardFinder::findInternal(robot_calibration_msgs::CalibrationData *
...
-    msg->observations[0].sensor_name = camera_sensor_name_;
-    msg->observations[0].features.resize(points_x_ * points_y_);
-    msg->observations[1].sensor_name = chain_sensor_name_;
-    msg->observations[1].features.resize(points_x_ * points_y_);
+    // Handle multiple finders
+
+    int idx_cam = -1;
+    int idx_chain = -1;
+
+    if (msg->observations.size() == 0)
```
Comment (mikeferguson) [inline robot_calibration/src/checkerboard_finder.cpp]: we probably don't need this block -- each finder should be the exclusive creator for each sensor, right? thus, we should always just append two new observations

File: robot_calibration/src/ground_plane_finder.cpp [hunk 9]
```diff
@@ -131,16 +131,46 @@ bool GroundPlaneFinder::find(robot_calibration_msgs::CalibrationData * msg)
...
-  msg->observations.resize(2);
-  msg->observations[0].sensor_name = camera_sensor_name_;
-  msg->observations[0].features.resize(points_total);
-  msg->observations[1].sensor_name = chain_sensor_name_;
-  msg->observations[1].features.resize(points_total);
+  int idx_cam = -1;
+  int idx_chain = -1;
 
-  size_t step = cloud_.width/points_total;
+  if (msg->observations.size() == 0)
```
Comment (mikeferguson) [inline robot_calibration/src/ground_plane_finder.cpp]: we probably don't need this block -- each finder should be the exclusive creator for each sensor, right? thus, we should always just append two new observations

File: robot_calibration/src/led_finder.cpp [hunk 10]
```diff
@@ -247,9 +249,40 @@ bool LedFinder::find(robot_calibration_msgs::CalibrationData * msg)
...
 
   // Export results
-  msg->observations.resize(2);
-  msg->observations[0].sensor_name = camera_sensor_name_;
-  msg->observations[1].sensor_name = chain_sensor_name_;
+  // Handle multiple finders
+  int idx_cam = -1;
+  int idx_chain = -1;
+
+  if (msg->observations.size() == 0)
```
Comment (mikeferguson) [inline robot_calibration/src/led_finder.cpp]: we probably don't need this block -- each finder should be the exclusive creator for each sensor, right? thus, we should always just append two new observations

**Comment by narora1** [comment]
fixed!

File: robot_calibration/src/led_finder.cpp [hunk 11]
```diff
@@ -21,6 +21,8 @@
 #include <robot_calibration/capture/led_finder.h>
 #include <sensor_msgs/point_cloud2_iterator.h>
 #include <sensor_msgs/image_encodings.h>
+#include <opencv2/core/core.hpp>
+#include <opencv2/core/types_c.h>
```
Comment (mikeferguson) [inline robot_calibration/src/led_finder.cpp]: these includes should not be needed, given that we haven't added anything related to them?

File: robot_calibration/src/optimizer.cpp [hunk 12]
```diff
@@ -146,32 +144,42 @@ int Optimizer::optimize(OptimizationParams& params,
...
-          models_[arm_name],
-          offsets_.get(), data[i]);
+            dynamic_cast<Camera3dModel*>(models_[camera_name]),
+            models_[arm_name],
+            offsets_.get(), data[i]);
+
+        int index = -1;
+        for (size_t k =0; k < data[i].observations.size() ; k++)
+        {
+            if ( data[i].observations[k].sensor_name == camera_name)
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: this is still indented wrong, and should not have a space after opening parenthesis

File: robot_calibration/src/optimizer.cpp [hunk 13]
```diff
@@ -146,32 +144,42 @@ int Optimizer::optimize(OptimizationParams& params,
...
-          for (size_t k = 0; k < data[i].observations[0].features.size(); ++k)
+          for (size_t k = 0; k < data[i].observations[index].features.size(); ++k)
             std::cout << "  " << std::setw(10) << std::fixed << residuals[(3*k + 2)];
           std::cout << std::endl << std::endl;
         }
 
         problem->AddResidualBlock(cost,
-                                   NULL /* squared loss */,
-                                   free_params);
+                                  NULL,//  squared loss
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: should be 2 spaces in front of //

File: robot_calibration/src/optimizer.cpp [hunk 14]
```diff
@@ -188,16 +196,26 @@ int Optimizer::optimize(OptimizationParams& params,
...
+        int index = -1;
+        for (size_t k =0; k < data[i].observations.size() ; k++)
+        {
+          if ( data[i].observations[k].sensor_name == camera_name)
+          {
+            index = k;
+            break;
+          }
+        }
+
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: this is missing a check for index = -1

File: robot_calibration/src/optimizer.cpp [hunk 15]
```diff
@@ -15,8 +15,6 @@
  * limitations under the License.
  */
 
-// Author: Michael Ferguson
-
```
Comment (mikeferguson) [inline robot_calibration/src/optimizer.cpp]: huh?

File: robot_calibration/src/led_finder.cpp [hunk 16]
```diff
@@ -247,9 +249,13 @@ bool LedFinder::find(robot_calibration_msgs::CalibrationData * msg)
   sensor_msgs::PointCloud2Iterator<float> iter_cloud(cloud, "x");
 
   // Export results
-  msg->observations.resize(2);
-  msg->observations[0].sensor_name = camera_sensor_name_;
-  msg->observations[1].sensor_name = chain_sensor_name_;
+  // Handle multiple finders
```
Comment (mikeferguson) [inline robot_calibration/src/led_finder.cpp]: we don't really need these "handle multiple finders" comments

File: robot_calibration/src/ground_plane_finder.cpp [hunk 17]
```diff
@@ -131,16 +131,19 @@ bool GroundPlaneFinder::find(robot_calibration_msgs::CalibrationData * msg)
...
+  msg->observations[idx_chain].sensor_name = chain_sensor_name_;
+  
+  msg->observations[idx_cam].features.resize(points_total);
+  msg->observations[idx_chain].features.resize(points_total);
+
+  size_t step = cloud_.width/(1.5 * points_total);
   size_t k = 0;
 
-  for (size_t i = step; i < cloud_.width; i +=step)
+  for (size_t i = step; i < cloud_.width/1.5; i +=step)
```
Comment (mikeferguson) [inline robot_calibration/src/ground_plane_finder.cpp]: what is up with this 1.5? this seems like a change in performance, we should probably put it in a separate commit

File: robot_calibration/src/ground_plane_finder.cpp [hunk 18]
```diff
@@ -131,16 +131,19 @@ bool GroundPlaneFinder::find(robot_calibration_msgs::CalibrationData * msg)
...
+  int idx_cam = msg->observations.size() + 0;
+  int idx_chain = msg->observations.size() + 1;
+  msg->observations.resize(msg->observations.size() + 2);
+  msg->observations[idx_cam].sensor_name = camera_sensor_name_;
+  msg->observations[idx_chain].sensor_name = chain_sensor_name_;
+  
+  msg->observations[idx_cam].features.resize(points_total);
+  msg->observations[idx_chain].features.resize(points_total);
+
+  size_t step = cloud_.width/(1.5 * points_total);
```
Comment (mikeferguson) [inline robot_calibration/src/ground_plane_finder.cpp]: this also has that 1.5....

File: robot_calibration/src/checkerboard_finder.cpp [hunk 19]
```diff
@@ -172,11 +172,17 @@ bool CheckerboardFinder::findInternal(robot_calibration_msgs::CalibrationData *
     sensor_msgs::PointCloud2Iterator<float> iter_cloud(cloud_, "x");
 
     // Set msg size
-    msg->observations.resize(2);
-    msg->observations[0].sensor_name = camera_sensor_name_;
-    msg->observations[0].features.resize(points_x_ * points_y_);
-    msg->observations[1].sensor_name = chain_sensor_name_;
-    msg->observations[1].features.resize(points_x_ * points_y_);
+    // Handle multiple finders
+
```
Comment (mikeferguson) [inline robot_calibration/src/checkerboard_finder.cpp]: handle multiple finders and blank line can be removed

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 20]
```diff
@@ -66,11 +66,21 @@ struct GroundPlaneError
...
+    int index = -1;
+    for (size_t k =0; k < data.observations.size() ; k++)
+    {
+      if ( data.observations[k].sensor_name == "cameraground")
+      {
+        index = k;
+        break;
+      }
+    }
+
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: also missing the index = -1 check here

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h [hunk 21]
```diff
@@ -56,7 +56,7 @@ struct GroundPlaneError
     // Compute residuals
     for (size_t i = 0; i < camera_pts.size() ; ++i)
     {
-      residuals[i] = camera_pts[i].point.z - z_;  // if camera_pts is in base frame
+      residuals[i] = (camera_pts[i].point.z - z_);  // if camera_pts is in base frame
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: these brackets are an unnecessary change

File: robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h [hunk 22]
```diff
@@ -106,11 +106,27 @@ struct Camera3dToArmError
                                      CalibrationOffsetParser* offsets,
                                      robot_calibration_msgs::CalibrationData& data)
   {
+    int index = -1;
+    for (size_t k = 0; k < data.observations.size() ; k++)
+    {
+      if ( data.observations[k].sensor_name == "camera")
```
Comment (mikeferguson) [inline robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h]: "camera" should probably not be hard coded -- but it looks like we don't currently expose the camera model name -- so we probably need to adjust that API

File: robot_calibration/include/robot_calibration/ceres/ground_plane_error.h (hunk 20 above)
Comment (narora1) [inline robot_calibration/include/robot_calibration/ceres/ground_plane_error.h]: I added this check but this isn't collapsing.

File: robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h (hunk 22 above)
Comment (narora1) [inline robot_calibration/include/robot_calibration/ceres/camera3d_to_arm_error.h]: will probably do this once I get the depth finder working?

File: robot_calibration/src/optimizer.cpp (hunk 15 above)
Comment (narora1) [inline robot_calibration/src/optimizer.cpp]: no idea how this happened :/
